tasks/EchoesToFrameConverterTask.cpp

```diff
@@ -0,0 +1,159 @@
+#include <base/samples/RigidBodyState.hpp>
+#include <math.h>
+#include <opencv2/imgproc/imgproc.hpp>
+
+using namespace radar_base;
+using namespace cv;
+using namespace base::samples::frame;
+EchoesToFrameConverterTask::EchoesToFrameConverterTask(std::string const& name)
+    : EchoesToFrameConverterTaskBase(name)
+{
+}
+
+EchoesToFrameConverterTask::~EchoesToFrameConverterTask()
+{
+}
+
+/// The following lines are template definitions for the various state machine
+// hooks defined by Orocos::RTT. See EchoesToFrameConverterTask.hpp for more detailed
+// documentation about them.
+
+bool EchoesToFrameConverterTask::configureHook()
+{
+    if (!EchoesToFrameConverterTaskBase::configureHook())
+        return false;
+    configureOutput();
+    updateLookUpTable();
+    return true;
+}
+bool EchoesToFrameConverterTask::startHook()
+{
+    if (!EchoesToFrameConverterTaskBase::startHook())
+        return false;
+    return true;
+}
+void EchoesToFrameConverterTask::updateHook()
+{
+    EchoesToFrameConverterTaskBase::updateHook();
+
+    base::Angle yaw_correction = base::Angle::fromRad(0);
+    base::samples::RigidBodyState sensor2ref_pose;
+    if (_sensor2ref_pose.read(sensor2ref_pose) != RTT::NoData) {
+        yaw_correction = base::Angle::fromRad(sensor2ref_pose.getYaw());
+    }
+
+    Radar radar_echo;
+    while (_echo.read(radar_echo, false) == RTT::NewData) {
+        if (m_current_sweep_size != radar_echo.sweep_length ||
+            m_current_num_angles != (2 * M_PI / radar_echo.step_angle.getRad()) ||
+            m_current_range != radar_echo.range) {
+            m_current_sweep_size = radar_echo.sweep_length;
+            m_current_num_angles = 2 * M_PI / radar_echo.step_angle.getRad();
+            m_current_range = radar_echo.range;
+            updateLookUpTable();
+        }
+
+        addEchoesToFrame(radar_echo, yaw_correction);
+    }
+    if (base::Time::now() - m_last_sample > _export_config.get().time_between_frames) {
+        publishFrame();
+    }
+}
+void EchoesToFrameConverterTask::errorHook()
+{
+    EchoesToFrameConverterTaskBase::errorHook();
+}
+void EchoesToFrameConverterTask::stopHook()
+{
+    EchoesToFrameConverterTaskBase::stopHook();
+}
+void EchoesToFrameConverterTask::cleanupHook()
+{
+    EchoesToFrameConverterTaskBase::cleanupHook();
+}
+
+void EchoesToFrameConverterTask::updateLookUpTable()
+{
+    LOG_INFO_S << "Updating LookUpTable";
+    auto config = _export_config.get();
+    m_lut.reset(new EchoToImageLUT(m_current_num_angles,
+        m_current_sweep_size,
+        config.beam_width,
+        config.output_image_size));
+    m_echoes = std::vector<uint8_t>(m_current_sweep_size * m_current_num_angles, 0);
+}
+
+void EchoesToFrameConverterTask::addEchoesToFrame(Radar const& echo,
+    base::Angle yaw_correction)
+{
+    LOG_INFO_S << "Adding echoes...";
+    double angle =
+        base::Angle::normalizeRad(echo.start_heading.getRad() + yaw_correction.getRad());
```

Comment:
`(echo.start_heading + yaw_correction).getRad()`, which removes the  need for normalizeRad (and is clearer)